File: Practice/oo.lua
--使用原型的方式面向对象

Base = {}
function Base:func2()
	print("Base", self.name)
end

A = {}
setmetatable(A, A)
A.__index = Base
function A:func ()
	print("A", self.name)
end

a = {}
a.name = "linn"
setmetatable(a, a)
a.__index = A

a:func()
a:func2()

b = {}
b.name = "base"
setmetatable(b, b)
b.__index = Base

b:func2()
b:func()

File: Practice/oo.js
function Base(n){
	this.name = n;
}

Base.prototype.func = function() {
	console.log("In Base func, " + this.name);
};

var o = new Base("linn");
o.func();



function Tmp(n){
	this.name = n;
}

TmpProto = {
	func: function (){
		console.log("In TmpProto func, " + this.name);
	},

	func2: function(callback){
		callback(this.name);
	}
};

Tmp.prototype = TmpProto;

var t = new Tmp("linn");
t.func();

function callback(n){
	console.log("In callback " + n);
	console.log("This " + this);
}

t.func2(callback);